### ISSUE TITLE
ScoreListener reports zero loss while training in SameDiff

### ISSUE THREAD

**Issue by CompilerCrash** [deeplearning4j/deeplearning4j#9684]
#### Issue Description

When I fit my SameDiff graph, the reported loss (as per ScoreListener and LossCurve) is constantly zero although the learning succeeds. This happened with different loss functions (for regression and classification). Below is a small example that should reproduce the situation.

- expected behavior: The reported loss should be greater than zero and decrease during the training.
- encountered behavior: The reported loss is constantly zero.
```
SameDiff sd = SameDiff.create();

int batchSize = 4;
int seqLength = 8;

SDVariable features = sd.squeeze(sd.placeHolder("features", FLOAT, batchSize, 1, seqLength), 1);
SDVariable labels = sd.squeeze(sd.placeHolder("labels", FLOAT, batchSize, 1, seqLength), 1);
SDVariable bias = sd.var("bias", new ZeroInitScheme('c'), FLOAT, seqLength);
SDVariable predictions = features.add("predictions", bias);
sd.loss.meanSquaredError("loss", labels, predictions, null);

TrainingConfig config = new TrainingConfig.Builder()
        .updater(new Adam(0.1))
        .dataSetFeatureMapping("features")
        .dataSetLabelMapping("labels")
        .build();
sd.setTrainingConfig(config);

// Task: add 1 to the inputs
SequenceRecordReader featureReader = new CSVLineSequenceRecordReader();
featureReader.initialize(new FileSplit(new File("src/main/resources/features.csv")));
SequenceRecordReader labelReader = new CSVLineSequenceRecordReader();
labelReader.initialize(new FileSplit(new File("src/main/resources/labels.csv")));
DataSetIterator iterator = new SequenceRecordReaderDataSetIterator(featureReader, labelReader, batchSize, 0, true);

// ScoreListener will consistently report a loss of 0
History hist = sd.fit(iterator, 10, new ScoreListener(1));

// The recorded loss values are also constantly 0
LossCurve curve = hist.lossCurve();
System.out.println("Loss curve:");
System.out.println(curve.getLossNames());
System.out.println(curve.getLossValues());

// However, the loss calculated here is > 0
Map<String, INDArray> map = sd.output(iterator, "loss", "bias");
System.out.println("Final loss: " + map.get("loss"));
System.out.println("Final bias: " + map.get("bias"));
```

#### Version Information

* Deeplearning4j version: 1.0.0-SNAPSHOT
* Platform information (OS, etc): Linux Mint 20.3
* CUDA version: not used
* NVIDIA driver version: not in use

#### Additional Information

* Shortened log:
```
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 4
[main] INFO org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Binary level Generic x86 optimization level Generic x86
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for OpenMP BLAS: 2
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [4]; Memory: [1,9GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
[main] INFO org.nd4j.linalg.cpu.nativecpu.CpuBackend - Backend build information:
 GCC: "7.5.0"
STD version: 201103L
DEFAULT_ENGINE: samediff::ENGINE_CPU
HAVE_FLATBUFFERS
HAVE_OPENBLAS
[main] WARN org.nd4j.imports.converters.ImportClassMapping - Duplicate TF op mapping found for op Pow: org.nd4j.linalg.api.ops.impl.scalar.Pow vs org.nd4j.linalg.api.ops.impl.transforms.custom.Pow
[main] WARN org.nd4j.imports.converters.ImportClassMapping - Duplicate TF op mapping found for op FloorMod: org.nd4j.linalg.api.ops.impl.transforms.pairwise.arithmetic.FModOp vs org.nd4j.linalg.api.ops.impl.transforms.pairwise.arithmetic.FloorModOp
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 0 complete on iteration 0 - 1 batches (4 examples) in 74 ms - 13,51 batches/sec, 54,05 examples/sec, 0 ms ETL time
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 1: 0,00000E0
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 1 complete on iteration 1 - 1 batches (4 examples) in 28 ms - 35,71 batches/sec, 142,86 examples/sec, 0 ms ETL time
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 2, iteration 2: 0,00000E0
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 2 complete on iteration 2 - 1 batches (4 examples) in 14 ms - 71,43 batches/sec, 285,71 examples/sec, 0 ms ETL time
...
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 8, iteration 8: 0,00000E0
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 8 complete on iteration 8 - 1 batches (4 examples) in 9 ms - 111,11 batches/sec, 444,44 examples/sec, 0 ms ETL time
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 9, iteration 9: 0,00000E0
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 9 complete on iteration 9 - 1 batches (4 examples) in 11 ms - 90,91 batches/sec, 363,64 examples/sec, 0 ms ETL time
[main] INFO org.nd4j.autodiff.samediff.SameDiff - Creating getInferenceFactory().create for thread 1
Loss curve:
[loss]
[[0], 
 [0], 
 ...
 [0], 
 [0]]
Final loss: 0.0058
Final bias: [    0.9237,    0.9237,    0.9237,    0.9237,    0.9237,    0.9237,    0.9237,    0.9237]
```
* pom.xml file see [here](https://github.com/CompilerCrash/DL4J-Bug-Reproduction/blob/master/pom.xml)
* Full repository [here](https://github.com/CompilerCrash/DL4J-Bug-Reproduction)
* I first thought that this was issue #8542 but that one is already fixed.

#### Contributing

Sorry, no.

**Comment by agibsonccc** [comment]
This is a dup of: https://github.com/eclipse/deeplearning4j/issues/9655 - closing. Please feel free to let me know if something's still not working.

**Comment by CompilerCrash** [comment]
@agibsonccc Oh, I overlooked that issue. I tried the code provided in its description and can confirm that the bug was present in M2 and that it is now fixed in SNAPSHOT.

However, using the very same SNAPSHOT, my above code is still not working. So, as far as I can tell, this issue is not a duplicate. Can you take a second look at it, please?

---

~~Side note: In the current SNAPSHOT, printing INDArrays gives e. g. `java.lang.IllegalStateException: Invalid interval: Interval(b=0,e=101,s=1,inclusive) on array with shape [100, 1]`. You might also want to take a look at this.~~ EDIT: This is fixed now.

**Comment by CompilerCrash** [comment]
@agibsonccc I double-checked everything, but I cannot find the error on my side (I would be happy if it was just that). I think it's not a duplicate. It would be great if you could take a look.

**Comment by CompilerCrash** [comment]
I was wrong about the training success. Given the right hyperparameters, this part works. I updated the issue description accordingly so that it is not misleading. However, the reported loss (which consequently cannot be the "real" loss) is still constantly zero. Maybe there is a bug in recording the loss values?

**Comment by CompilerCrash** [comment]
To future readers: The same problem was brought up again in issue #9806, which even received some fix, but still the loss reported by `ScoreListener` and `LossCurve` is constantly zero in the given examples.

Fortunately, I found an easy workaround. You need to manually pass the names of the loss variables to your `TrainingConfig`. For example:
```
TrainingConfig config = new TrainingConfig.Builder()
	// ...
	.lossVariables(List.of("my_loss"))
	// ...
	.build();

// or alternatively:
config.setLossVariables(List.of("my_loss"));
```

**Comment by agibsonccc** [comment]
@CompilerCrash mind filing a separate issue?
